src/config/config.ts

```diff
@@ -30,6 +30,35 @@ function isConfig(o: any): o is IConfig {
   return o && Boolean(o._base)
 }
 
+class PermutationIndex extends Map<string, Set<string>> {
+  private validPermutations = new Map<string, string>()
+
+  public add(permutation: string, commandId: string): void {
+    this.validPermutations.set(permutation, commandId)
+    for (const part of collectUsableIds([permutation])) {
+      if (this.has(part)) {
+        this.set(part, this.get(part).add(commandId))
+      } else {
+        this.set(part, new Set([commandId]))
+      }
+    }
+  }
+
+  public get(key: string): Set<string> {
+    return super.get(key) ?? new Set()
+  }
+
+  public getValid(): string[]
+  public getValid(key: string): string | undefined
+  public getValid(key?: string): string | string[] | undefined {
+    return key ? this.validPermutations.get(key) : [...this.validPermutations.keys()]
+  }
```

Comment:
Can we split this into two functions to make the code more readable/understandable?
Something like:
```suggestion
  public getValid(key: string): string | undefined {
    return this.validPermutations.get(key);
  }
  
  public getAllValid(): string[] {
    return [...this.validPermutations.keys()];
  }
```